core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -447,6 +446,9 @@ private IndexShardState changeState(IndexShardState newState, String reason) {
 
     public Engine.Index prepareIndexOnReplica(SourceToParse source, long version, VersionType versionType) {
         try {
+            if (shardRouting.primary() && shardRouting.isRelocationTarget() == false) {
+                throw new IllegalIndexShardStateException(shardId, state, "shard is not a replica");
```

Comment:
I think this check is wrong. When we have relocation going on and relocation source is marked as relocated (i.e. we call executeRemotely in TransportReplicationAction), then we have primary relocation target replicating back to primary relocation source (see also ReplicationPhase).